keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,931 @@
+the Proposed state makes the Kubelet the only writer to that field, but leaves a
+gap in knowing whether the latest resize has been acknowledged by the Kubelet. 
+The changes proposed in this KEP resolve this gap.
+
+The ResizeStatus is used as an example here, but in practice this issue can be 
+generally found in any type of pod update.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Provide a general solution for the pod status to express which pod update is 
+currently being reflected.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Address more specific issues with any particular pod status field(s). 
+* Expand the set of mutable fields.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes
+
+#### Generation and ObservedGeneration
+
+The `metadata.generation` field is currently unused on pods, but we can start 
+leveraging it to help track which pod state is currently being reflected in the 
+pod status. [Generation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) is documented as "a sequence 
+number representing a specific generation of the desired state. Set by the 
+system and monotonically increasing, per-resource." For consistency, pod 
+Generation will be incremented whenever the pod has changes that the kubelet 
+needs to actuate. 
+
+A new field ObservedGeneration will be added to the pod status. 
+Kubelet will set this to communicate which pod state is being expressed in the 
+current pod status. This is analogous to the ObservedGeneration that exists
+in other resources' statuses such as [StatefulSets](https://github.com/kubernetes/api/blob/9e7d345b161c12a3056efa88e285b3ef68450c54/apps/v1/types.go#L274).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+No risks have been identified.
```

Comment:
If a custom-set generation is honored at pod creation time today, one possible risk is that an external client depends on being able to set metadata.generation on pod create, and depends on it never changing, and uses that in its own reconciliation logic somehow.

No *in-tree* controllers do this from what I can tell ... they have their own solutions for this (for example, DaemonSetController puts the daemonset template generation in a label on the pod, then looks for that label value), but I could imagine an external controller spawning pods with metadata.generation=$parentControllerObj.metadata.generation, then comparing pod.metadata.generation to $parentControllerObj.metadata.generation to see if those pods are "current" for the controller.

Reply:
> If a custom-set generation is honored at pod creation time today, one possible risk is that an external client depends on being able to set metadata.generation on pod create, and depends on it never changing, and uses that in its own reconciliation logic somehow.

This thought also crossed my mind (I should have written it down), but it seemed a bad practice for an external client to set and depend on a custom value for a field documented as "set by the system" that I think I wrote it off in my mind as unlikely. But you are right, it is a possible risk so I'll document it in the KEP

Reply:
from https://github.com/kubernetes/enhancements/pull/5068#discussion_r1927652561

> Oh sorry I missed your question about seeing what happens on graceful delete of a pod. Just tried it - it does actually bump the pod generation today so we can keep that behavior

The fact that generation was already getting incremented on graceful delete is actually incredibly helpful in de-risking the chances that an external controller was doing something fragile in expecting to set generation to a custom value and have it never change 👍